### PR TITLE
Implemeted SQL function center() that calculates the gravity center(centroid) from array of points

### PR DESCRIPTION
**Pull request by stepan-neretin7**

### CODE REVIEW

**Comment by stepan-neretin7** [comment]
@esabol @vitcpp please review it.

File: Makefile [hunk 1]
```diff
@@ -31,7 +31,8 @@ DATA_built  = $(RELEASE_SQL) \
 
 DOCS        = README.pg_sphere COPYRIGHT.pg_sphere
 REGRESS     = init tables points euler circle line ellipse poly path box index \
-              contains_ops contains_ops_compat bounding_box_gist gnomo epochprop
+              contains_ops contains_ops_compat bounding_box_gist gnomo epochprop\
```
Comment (esabol) [inline Makefile:34]: There should be a space between `echoprop` and the backslash on this line.

File: doc/operators.sgm [hunk 2]
```diff
@@ -393,7 +393,7 @@
                 The center operator <literal>@@</literal> is a non-boolean
                 unary operator returning the center of an object.  In the
                 current implementation of <application>pgSphere</application>,
-                only centers of circles and ellipses are supported. Instead
+                only centers of circles, ellipses, point, array of points are supported. Instead
```
Comment (esabol) [inline doc/operators.sgm:396]: I think it should be `points, arrays of points` to be grammatically correct here. I think this line needs is too long, and it needs to be word-wrapped after `are`, along with the next couple of lines where needed.

File: pgs_point.sql.in [hunk 3]
```diff
@@ -167,3 +167,33 @@ CREATE OPERATOR  <-> (
 COMMENT ON OPERATOR <-> (spoint, spoint) IS
   'distance between spherical points';
 
+CREATE FUNCTION center(dots_vector SPoint[])
```
Comment (esabol) [inline pgs_point.sql.in:170]: I thought we had settled on naming this function `centroid`? Why was `center` chosen?

I see it referred to as "centroid" at these URLs:

https://stackoverflow.com/a/38201499/5153779
https://gis.stackexchange.com/questions/43505/calculating-a-spherical-polygon-centroid

However, "center" is used here:
https://github.com/chrisveness/geodesy/blob/8f4ef33d/latlon-nvector-spherical.js#L783

So I'm not sure which term is correct.

Does your implementation give the same results as the Python and JavaScript implementations at https://stackoverflow.com/questions/19897187#answer-38201499 and https://github.com/chrisveness/geodesy/blob/8f4ef33d/latlon-nvector-spherical.js#L783 ? If not, why not?

**Review by esabol — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (stepan-neretin7) [reply Makefile:34]: resolved

File: doc/operators.sgm (hunk 2 above)
Comment (stepan-neretin7) [reply doc/operators.sgm:396]: resolved

File: pgs_point.sql.in (hunk 3 above)
Comment (stepan-neretin7) [reply pgs_point.sql.in:170]: @vitcpp and I discussed the name "centroid" earlier, but in the end we came to the conclusion that the best name is the center for the sql function to match the interface with the operator @@

Comment (vitcpp) [reply pgs_point.sql.in:170]: @esabol The main reason to choose 'center' function because such function already exists. It returns the center of circle or ellipse and it is intended to be expanded to other object types. In the doc it is described as center of an object. It was decided to overload this function and allow to pass array of points and the point itself (for interface completeness). It is why we decided to put on display PR with 'center' function. Furthermore, we may utilize operator @@. My opinion is that center and centroid are interchangeable in most cases. So, our idea is to reuse center and to avoid introducing new terms.

Comment (esabol) [reply pgs_point.sql.in:170]: OK, that makes sense (the `center` name) then, but I'd still like an answer to this:

> Does your implementation give the same results as the Python and JavaScript implementations at https://stackoverflow.com/questions/19897187#answer-38201499 and https://github.com/chrisveness/geodesy/blob/8f4ef33d/latlon-nvector-spherical.js#L783 ? If not, why not?

File: doc/operators.sgm [hunk 4]
```diff
@@ -393,8 +393,10 @@
                 The center operator <literal>@@</literal> is a non-boolean
                 unary operator returning the center of an object.  In the
                 current implementation of <application>pgSphere</application>,
-                only centers of circles and ellipses are supported. Instead
-                of using the operator, you can use the function
+                only centers of circles, ellipses, points, array of points are
```
Comment (esabol) [inline doc/operators.sgm:396]: Should be `arrays of points`.

**Review by esabol — CHANGES_REQUESTED** [review]

File: doc/operators.sgm (hunk 4 above)
Comment (stepan-neretin7) [reply doc/operators.sgm:396]: ???
I also add center function and operator for single spoint for completeness interface

Comment (esabol) [reply doc/operators.sgm:396]: Let me rephrase: Change "array of points" on this line to "arrays of points" to make this sentence grammatically correct.

File: pgs_point.sql.in (hunk 3 above)
Comment (stepan-neretin7) [reply pgs_point.sql.in:170]: @esabol But this is function, which 
> Calculates the centre of a spherical **polygon** where **the sides of the polygon are great circle arcs** joining the vertices.
In my task, I have an arbitrary set of points. That is why I have now tested and on any two input points their function gives 0, because the polygon starts with three points

File: doc/operators.sgm (hunk 4 above)
Comment (stepan-neretin7) [reply doc/operators.sgm]: resolved

File: src/point.h [hunk 5]
```diff
@@ -15,6 +15,11 @@ typedef struct
 	float8	lat;	/* latitude value in radians */
 } SPoint;
 
+/*
+ * For the correctness of the interface, the function returns the spoint itself
+*/
```
Comment (esabol) [inline src/point.h:20]: Trivial: Please insert a space before the asterisk on this line. The asterisks should line up vertically.

**Review by esabol — CHANGES_REQUESTED** [review]

File: pgs_point.sql.in (hunk 3 above)
Comment (esabol) [reply pgs_point.sql.in:170]: Are you saying that what your `center()` function calculates is different from what is being calculated by the function at https://stackoverflow.com/questions/19897187#answer-38201499 using the Stokes theorem?

File: src/point.h (hunk 5 above)
Comment (stepan-neretin7) [reply src/point.h]: resolved

File: pgs_point.sql.in (hunk 3 above)
Comment (stepan-neretin7) [reply pgs_point.sql.in:170]: @esabol yes, because it specified on **polygon** points

Comment (esabol) [reply pgs_point.sql.in:170]: Are you saying your function works on points not on the surface of the same sphere?

Comment (stepan-neretin7) [reply pgs_point.sql.in:170]: @esabol hmm, I think yes, it is generally about some points
The logic itself is not tied to the sphere, the general definition of the center of mass was used.

Comment (stepan-neretin7) [reply pgs_point.sql.in:170]: @esabol And I think my algorithm is quite attached to reality, because, for example, for two opposite poles, it gives the center of the sphere.

Comment (stepan-neretin7) [reply pgs_point.sql.in:170]: @esabol Sorry... Give me time, then I'll try to figure out why the answers are completely different

Comment (vitcpp) [reply pgs_point.sql.in:170]: Dear Colleagues, it was intended that the proposed center function calculates the geometrical center of a set of points (particles) that do not form a figure. Such function may be useful in some astronomical tasks when we have to deal with stellar clusters or with other groups of astronomical objects which geometrical center on the sphere we'd like to find. When we want to find the center of a figure we definitely should use the Stokes theorem as it was implemented in the referenced JS library (center(spoly) should be implemented that way).  I guess it is why @stepan-neretin7 sees some differences when trying to compare the proposed function against the function to calculate polygon center from the referenced JS library. Adding some more points on a polygon edge doesn't change the shape of the polygon. But adding more points change the mass center of the set of particles (points). There is another question - is it reasonable to implement such function? I think yes. Probably, the name center is best suited to calculate the center of a figure than to calculate the mass center of a set of points (may be come back to centroid - ?). I'm not sure about having such similar function in the JS library. If yes, it is reasonable to compare our function against its counterpart.

Comment (esabol) [reply pgs_point.sql.in:170]: OK with me.

**Review by esabol — APPROVED** [review]

**Comment by vitcpp** [comment]
I have some concerns related to the cases when cartesian geometric center lies in the center of the sphere. The following situations are possible:

- The center on the sphere is absent. I guess in this case the function returns (0,0).

- There are two possible centers on the sphere when two points are locates on the opposite sides of a meridian. But in this case due to rounding issues in most situations out calculated center in 3D will be near the center of the sphere but not exactly there. It affects which point will be choosen as the result and may produce unstable results I guess.

**Comment by stepan-neretin7** [comment]
Other popular libraries, except javascript, do not even allow you to check which center of mass they give out for two antipodal points, because they write that a minimum of 6 points is needed to create a polygon. And for some reason the Javascript library writes (0,0), which is fundamentally wrong

**Comment by vitcpp** [comment]
We can't compare the function for calculation the centroid of a set of independent points with the calculation of a polygon figure. The set of independent points does not form a figure. We should compare the equivalent function in other spherical geometry libraries if such function exist. The only unresolved question - what to return when the points are on opposite sides. There are two cases: either we have two result points or there are no points at all (when the calculated center in 3D is near the center of the sphere).

File: pgs_point.sql.in (hunk 3 above)
Comment (stepan-neretin7) [reply pgs_point.sql.in:170]: @vitcpp @esabol 
Dear colleagues, I want to revive the discussion of my pull-request. I have two questions that I would like to discuss? First; After all, we'd better leave such an implementation (the center of an arbitrary set of points, not a polygon.. If we add more points on the side of an imaginary polygon from points, then my implementation will shift to this side) or we will think in the direction of that implementation for polygons through the Stokes theorem
Second: When two points are at the poles: point(0,-pi()/2), point(0,pi()/2), then the middle point will be ~epsilon and it is incorrect to translate this into lat, lng, because such a point is not projected onto the sphere. What should I give out in this case? NULL? But let's say the case with the points 0,0 and 0, pi
The center here is any point on a large circle perpendicular to the line connecting the points
What is better to return here? NAN? NULL?

Comment (vitcpp) [reply pgs_point.sql.in:170]: Deal All, the final result of this operation is to find centroid of a set of points, not the center of a polygon. We do not find the center of a figure like polygon in this task. The averaged vector is pretty fine solution. There are two special cases. Case 1: we find the center of two opposite points. In this case, the equidistant place is a circle, not a point. The geometric 
 center in the Cartesian coordinate system is the center of the sphere. Case 2: when we have, for example, two points on poles and 4 points on equator with longitudes 0, 90, 180, 270 degrees. In this case the geometric center in 3D is (0,0,0) but there is no geometric center on the sphere. Case 3: we have all points on the same big circle. In this case, there are two points are equidistant. I'm not sure about other special cases. I guess, there are no other special cases.

This function may be useful to find the approximate center of a group of astronomical objects that are placed in a local area on the sphere: the center of a star cluster, the center of a selected group of objects, etc. I think the defined special cases are impractical; I'm not sure in what tasks it can happen. 

I propose this function to return NULL if 3D geometric center is (0,0,0) +- EPSILON. It means, the algorithm can't find centroid of the point set. I think, NULL is better because one can use ifull or coalesce to define some result in such cases.

**Comment by vitcpp** [comment]
There are some conceptual problems with centroid function. I propose to close it as it is not required right now. Later we may reopen it if neeed.

**Comment by vitcpp** [comment]
Closed as incomplete.
